2005-05-01  Paul Brook  <paul@codesourcery.com>

	* elflink.c (_bfd_elf_merge_symbol): Skip weak definitions if a strong
	definition has already been seen.

--- binutils-2.16.90.0.2/bfd/elflink.c.skip-weak-defs-if-strong-defs-available	2005-04-29 19:50:25.000000000 +0200
+++ binutils-2.16.90.0.2/bfd/elflink.c	2005-05-04 00:04:53.604698753 +0200
@@ -1078,6 +1078,13 @@ _bfd_elf_merge_symbol (bfd *abfd,
       || h->root.type == bfd_link_hash_undefined)
     *size_change_ok = TRUE;
 
+  /* Skip weak definitions of symbols that are already defined.  */
+  if (newdef && olddef && newweak && !oldweak)
+    {
+      *skip = TRUE;
+      return TRUE;
+    }
+
   /* NEWDYNCOMMON and OLDDYNCOMMON indicate whether the new or old
      symbol, respectively, appears to be a common symbol in a dynamic
      object.  If a symbol appears in an uninitialized section, and is
